src/main/java/ca/corefacility/bioinformatics/irida/ria/web/services/UIMetadataImportService.java

```diff
@@ -1,172 +1,241 @@
+		this.metadataTemplateService = metadataTemplateService;
+		this.projectControllerUtils = projectControllerUtils;
+		this.metadataFileImportService = metadataFileImportService;
 	}
 
 	/**
-	 * Parse metadata from an csv file.
+	 * Upload CSV or Excel file containing sample metadata and extract the headers.  The file is stored in the session until
+	 * the column that corresponds to a {@link Sample} identifier has been sent.
 	 *
-	 * @param inputStream The inputStream of the csv file.
-	 * @return {@link SampleMetadataStorage} contains the metadata from file.
-	 * @throws IOException thrown if the extension does not exist.
+	 * @param session   {@link HttpSession}
+	 * @param projectId {@link Long} identifier for the current {@link Project}
+	 * @param file      {@link MultipartFile} The csv or excel file containing the metadata.
+	 * @return {@link Map} of headers and rows from the csv or excel file for the user to select the header corresponding the
+	 * {@link Sample} identifier.
 	 */
-	public SampleMetadataStorage parseCSV(InputStream inputStream) throws IOException {
+	public SampleMetadataStorage createProjectSampleMetadata(HttpSession session, long projectId, MultipartFile file) {
+		// We want to return a list of the table headers back to the UI.
 		SampleMetadataStorage storage = new SampleMetadataStorage();
-
-		CSVParser parser = CSVParser.parse(inputStream, StandardCharsets.UTF_8,
-				CSVFormat.RFC4180.withFirstRecordAsHeader()
-						.withTrim()
-						.withIgnoreEmptyLines());
-		List<Map<String, String>> rows = new ArrayList<>();
-
-		// save headers
-		Map<String, Integer> headers_set = parser.getHeaderMap();
-		List<String> headers_list = new ArrayList<>(headers_set.keySet());
-		storage.saveHeaders(headers_list);
-
-		// save data
-		for (CSVRecord row : parser) {
-			Map<String, String> rowMap = new HashMap<>();
-			for (String key : row.toMap()
-					.keySet()) {
-				String value = row.toMap()
-						.get(key);
-				rowMap.put(key, value);
+		try (InputStream inputStream = file.getInputStream()) {
+			String filename = file.getOriginalFilename();
+			String extension = Files.getFileExtension(filename);
+
+			// Check the file type
+			switch (extension) {
+			case "csv":
+				storage = metadataFileImportService.parseCSV(projectId, inputStream);
+				break;
+			case "xlsx":
+			case "xls":
+				storage = metadataFileImportService.parseExcel(projectId, inputStream, extension);
+				break;
+			default:
+				// Should never reach here as the uploader limits to .csv, .xlsx and .xlx files.
+				throw new MetadataImportFileTypeNotSupportedError(extension);
 			}
-			rows.add(rowMap);
-		}
 
-		storage.saveRows(rows);
-		parser.close();
+		} catch (FileNotFoundException e) {
+			logger.debug("No file found for uploading an excel file of metadata.");
+		} catch (IOException e) {
+			logger.error("Error opening file" + file.getOriginalFilename());
+		}
 
+		session.setAttribute("pm-" + projectId, storage);
```

Comment:
Why are we setting this in the session? There's a potential for a lot of data to be stored. Or am I looking at it incorrectly?

Reply:
I don't know if I'm the best person to answer this question, lol. If I remember correctly, I think the consensus was to keep this as is. I think the concern was if we store it in the backend, it's a lot of reads & writes and we aren't really wanting to persist the files.

Reply:
Oh ok. That makes sense. We'll in that case it's good with me